src/base_object-inl.h

```diff
@@ -39,6 +39,12 @@ inline BaseObject::BaseObject(Environment* env, v8::Local<v8::Object> handle)
   // nullptr in case it's accessed by the user before construction is complete.
   if (handle->InternalFieldCount() > 0)
     handle->SetAlignedPointerInInternalField(0, nullptr);
+  env_->AddCleanupHook(DeleteMe, static_cast<void*>(this));
```

Comment:
will DeleteMe() somehow conflict with `delete data.GetParameter();` in `BaseObject::WeakCallback()`? should DeleteMe() call `ClearWeak()` ?
I didn't read through all of the code please forgive me if I ask stupid question. Right now seems two mechanisms for clean up, one is `v8::PersistentBase< T >::SetWeak()` which relays on V8 GC and one the env cleanup hook you add now. could them be merged into one?

Reply:
> will DeleteMe() somehow conflict with `delete data.GetParameter();` in `BaseObject::WeakCallback()`?

I don’t think so… when one of these mechanisms is called, the other gets disabled in the destructor. That should work, right?

> Should DeleteMe() call `ClearWeak()` ?

The `delete self;` call destroys the persistent handle, so I don’t think calling `ClearWeak()` on that handle before it would make a difference?

> I didn't read through all of the code please forgive me if I ask stupid question.

No no, please keep these questions (which are not stupid) coming :) It’s good to talk this through.

> could them be merged into one?

Hm – how would we do this? There isn’t much code being added here, so I’m not sure how to merge those…

Also, there’s a third way to cleanup a `BaseObject`: To `delete` it explicitly somewhere else, like we do e.g. for `HandleWrap`s. I agree that lifetime management in Node core is currently not the easiest thing to keep track off, though.